### PR TITLE
Make Parser & Tokenizer generic over Dialect, remove `test_utils` module

### PR DESCRIPTION
**Pull request by not-my-profile**
* rustc monomorphizes generics which is more efficient than
  trait object methods which are dispatched dynamically
    
* the Parser struct no longer has a lifetime, making it overall
   more convenient (e.g it now is Send, Sync & UnwindSafe)

For details see the commit messages.

If you like these changes, please merge them without squashing.

### CODE REVIEW

File: tests/test_utils.rs [hunk 1]
```diff
@@ -26,31 +21,79 @@ use alloc::{
...
+macro_rules! nest {
+    ($base:expr $(, $join:expr)*) => {
+        TableFactor::NestedJoin(Box::new(TableWithJoins {
+            relation: $base,
+            joins: vec![$(join($join)),*]
+        }))
+    };
+}
+
+pub trait Parse {
```
Comment (alamb) [inline tests/test_utils.rs:38]: I think you are using this trait so you can have a `dyn Parse` and dispatch dynamically at runtime what dialect is being used. This is likely something other users of the crate might want / might be using (dynamically pick a dialect).

What would you think about putting this trait in the main  crate (or perhaps even calling it `Parser` and changing the struct to `ParserImpl`)?

**Review by alamb — COMMENTED** [review]
Thank you @not-my-profile  -- this is a very nice contribution

I had a few questions related to test changes and the use of `PhantomData` but otherwise this looks like a nice change to me, even though it is backwards incompatible.

cc @Dandandan  @houqp @andygrove and @nickolay

**Comment by alamb** [comment]
Looks like there are some CI failures as well

File: src/lib.rs [hunk 2]
```diff
@@ -45,9 +43,3 @@ pub mod dialect;
 pub mod keywords;
 pub mod parser;
 pub mod tokenizer;
-
-#[doc(hidden)]
-// This is required to make utilities accessible by both the crate-internal
-// unit-tests and by the integration tests <https://stackoverflow.com/a/44541071/1026>
-// External users are not supposed to rely on this module.
-pub mod test_utils;
```
[the remark this answers is not in the document]
Comment (not-my-profile) [reply src/lib.rs:53]: It was `#[doc(hidden)]` and documented that "External users are not supposed to rely on this module." So I really don't see the problem.

File: src/parser.rs [hunk 3]
```diff
@@ -104,45 +105,49 @@ impl fmt::Display for ParserError {
 #[cfg(feature = "std")]
 impl std::error::Error for ParserError {}
 
-pub struct Parser<'a> {
+pub struct Parser<D: Dialect = GenericDialect> {
     tokens: Vec<Token>,
     /// The index of the first unprocessed token in `self.tokens`
     index: usize,
-    dialect: &'a dyn Dialect,
+    dialect: PhantomData<D>,
```
[the remark this answers is not in the document]
Comment (not-my-profile) [reply src/parser.rs:112]: rustc requires that generic type parameters of structs are actually used or it fails with a compile error telling you to use `PhantomData`. Note that `PhantomData` consumes no space it's just used by the compiler for the static analysis, see [core::marker::PhantomData](https://doc.rust-lang.org/core/marker/struct.PhantomData.html) and [PhantomData in the Nomicon](https://doc.rust-lang.org/nomicon/phantom-data.html).

File: tests/test_utils.rs (hunk 1 above)
Comment (not-my-profile) [reply tests/test_utils.rs:38]: I highly doubt that any API users actually need `Parser` trait objects. Note that with generics it's still possible to "dynamically" choose which dialect is being used at runtime, you don't need trait objects for that. As an example for that see the updated `examples/cli.rs`.

File: src/lib.rs [hunk 4]
```diff
@@ -21,14 +21,12 @@
...
 //!
-//! let dialect = GenericDialect {}; // or AnsiDialect
-//!
 //! let sql = "SELECT a, b, 123, myfunc(b) \
 //!            FROM table_1 \
 //!            WHERE a > b AND b < 100 \
 //!            ORDER BY a DESC, b";
 //!
-//! let ast = Parser::parse_sql(&dialect, sql).unwrap();
+//! let ast = Parser::<GenericDialect>::parse_sql(sql).unwrap();
```
[the remark this answers is not in the document]
Comment (not-my-profile) [reply src/lib.rs:29]: I'd prefer to improve the documentation in a followup PR.

**Comment by not-my-profile** [comment]
Hi Andrew, thanks for the review! I fixed the CI failure (it was just caused by `std::marker::PhantomData` not being available with `no_std`, I changed it to `core::marker::PhantomData`) and answered your comments :)

File: tests/test_utils.rs (hunk 1 above)
Comment (alamb) [reply tests/test_utils.rs:38]: > As an example for that see the updated examples/cli.rs.

Indeed -- I think in the case of `examples/cli.rs` the distinction is that the call to `parse<D>` is now also generic (the generics perolate back up)

**Comment by alamb** [comment]
Thanks again for this PR. I am closing it for now to clean up the PR list. Please reopen if that was a mistake.

**Pull request closed without merging**